flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskExecutionDecorationTest.java

```diff
@@ -0,0 +1,172 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.runtime.checkpoint.CheckpointMetaData;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.checkpoint.CheckpointType;
+import org.apache.flink.runtime.state.CheckpointStorageLocationReference;
+import org.apache.flink.runtime.util.FatalExitExceptionHandler;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.runtime.tasks.mailbox.Mail;
+import org.apache.flink.streaming.runtime.tasks.mailbox.MailboxDefaultAction;
+import org.apache.flink.util.function.ThrowingRunnable;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Verifies that {@link StreamTask} {@link StreamTaskActionExecutor decorates execution} of actions that potentially needs to be synchronized.
+ */
+public class StreamTaskExecutionDecorationTest {
+	private CountingStreamTaskActionExecutor decorator;
+	private StreamTask<Object, StreamOperator<Object>> task;
+
+	@Test
+	public void testAbortCheckpointOnBarrierIsDecorated() throws Exception {
+		task.abortCheckpointOnBarrier(1, null);
+		verify();
+	}
+
+	@Test
+	public void testTriggerCheckpointOnBarrierIsDecorated() throws Exception {
+		task.triggerCheckpointOnBarrier(new CheckpointMetaData(1, 2), new CheckpointOptions(CheckpointType.CHECKPOINT, new CheckpointStorageLocationReference(new byte[]{1})), null);
+		verify();
+	}
+
+	@Test
+	public void testTriggerCheckpointAsyncIsDecorated() throws Exception {
+		task.triggerCheckpointAsync(new CheckpointMetaData(1, 2), new CheckpointOptions(CheckpointType.CHECKPOINT, new CheckpointStorageLocationReference(new byte[]{1})), false);
+		new Thread(() -> {
+			try {
+				Thread.sleep(50);
+			} catch (InterruptedException ex) {
+				ex.printStackTrace();
+			} finally {
+				task.mailboxProcessor.allActionsCompleted();
+			}
+		}).start();
+		task.mailboxProcessor.runMailboxLoop();
+		verify();
+	}
+
+	@Test
+	public void testMailboxExecutorIsDecorated() throws Exception {
+		CountDownLatch latch = new CountDownLatch(1);
+		task.mailboxProcessor.getMainMailboxExecutor().asExecutor("test").execute(() -> {
+			try {
+				verify();
+			} finally {
+				latch.countDown();
+			}
+		});
+		new Thread(() -> {
+			try {
+				latch.await();
+				task.mailboxProcessor.allActionsCompleted();
+			} catch (Exception ex) {
+				ex.printStackTrace();
+			}
+		}).start();
+		task.mailboxProcessor.runMailboxLoop();
+	}
+
+	@Before
+	public void before() {
+		decorator = new CountingStreamTaskActionExecutor();
+		task = new StreamTask<Object, StreamOperator<Object>>(new StreamTaskTest.DeclineDummyEnvironment(), null, FatalExitExceptionHandler.INSTANCE, decorator) {
```

Comment:
Could you please check if `MockStreamTask` is an option?

Reply:
There is no `operatorChain` for the `MockStreamTask` and it shouldn't `processInput()`.

Reply:
what do you mean @rkhachatryan about `operatorChain`? `MockStreamTask` doesn't process any input.